plugins/openai/src/llm.ts

```diff
@@ -0,0 +1,669 @@
+    if (opts.apiKey === undefined) {
+      throw new Error('Telnyx API key is required, whether as an argument or as $TELNYX_API_KEY');
+    }
+
+    return new LLM({
+      model: 'meta-llama/Meta-Llama-3.1-70B-Instruct',
+      baseURL: 'https://api.telnyx.com/v2/ai',
+      ...opts,
+    });
+  }
+
+  chat({
+    chatCtx,
+    fncCtx,
+    temperature,
+    n,
+    parallelToolCalls,
+  }: {
+    chatCtx: llm.ChatContext;
+    fncCtx?: llm.FunctionContext | undefined;
+    temperature?: number | undefined;
+    n?: number | undefined;
+    parallelToolCalls?: boolean | undefined;
+  }): LLMStream {
+    temperature = temperature || this.#opts.temperature;
+
+    return new LLMStream(
+      this.#client,
+      chatCtx,
+      fncCtx,
+      this.#opts,
+      parallelToolCalls,
+      temperature,
+      n,
+    );
+  }
+}
+
+export class LLMStream extends llm.LLMStream {
+  #toolCallId?: string;
+  #fncName?: string;
+  #fncRawArguments?: string;
+  #client: OpenAI;
+  #logger = log();
+  #id = randomUUID();
+
+  constructor(
+    client: OpenAI,
+    chatCtx: llm.ChatContext,
+    fncCtx: llm.FunctionContext | undefined,
+    opts: LLMOptions,
+    parallelToolCalls?: boolean,
+    temperature?: number,
+    n?: number,
+  ) {
+    super(chatCtx, fncCtx);
+    this.#client = client;
+    this.#run(opts, n, parallelToolCalls, temperature);
+  }
+
+  async #run(opts: LLMOptions, n?: number, parallelToolCalls?: boolean, temperature?: number) {
+    const tools = this.fncCtx
+      ? Object.entries(this.fncCtx).map(([name, func]) => ({
+          type: 'function' as const,
+          function: {
+            name,
+            description: func.description,
+            parameters: llm.oaiParams(func.parameters),
+          },
+        }))
+      : undefined;
+
+    const stream = await this.#client.chat.completions.create({
+      model: opts.model,
+      user: opts.user,
+      n,
+      messages: await Promise.all(
+        this.chatCtx.messages.map(async (m) => await buildMessage(m, this.#id)),
+      ),
+      temperature: temperature || opts.temperature,
+      stream_options: { include_usage: true },
+      stream: true,
+      tools,
+      parallel_tool_calls: this.fncCtx && parallelToolCalls,
+    });
+
+    for await (const chunk of stream) {
+      for (const choice of chunk.choices) {
+        const chatChunk = this.#parseChoice(chunk.id, choice);
+        if (chatChunk) {
+          console.log(chatChunk.choices[0]);
+          this.queue.put(chatChunk);
+        }
+
+        if (chunk.usage) {
+          const usage = chunk.usage;
+          this.queue.put({
+            requestId: chunk.id,
+            choices: [],
+            usage: {
+              completionTokens: usage.completion_tokens,
+              promptTokens: usage.prompt_tokens,
+              totalTokens: usage.total_tokens,
+            },
+          });
+        }
+      }
+    }
+
+    this.queue.close();
```

Comment:
we should put the queue close inside a try.. finally. 
Otherwise it will never close the AsyncIterator if an error occured